File: packages/vscode/src/safety-interceptor.ts
// =============================================================================
// @codeforge/vscode — Safety Interceptor
// Warns before destructive commands are executed in the terminal
// =============================================================================

import * as vscode from "vscode";
import { CodeForgeRecorder } from "@codeforge/core";

/** Commands that are considered destructive */
const DESTRUCTIVE_PATTERNS: Array<{ pattern: RegExp; reason: string }> = [
  {
    pattern: /\brm\s+(-[rRfF]+\s+|-[a-zA-Z]*[rR][a-zA-Z]*\s+)/,
    reason: "This will permanently delete files and folders. There is no undo.",
  },
  {
    pattern: /\bgit\s+reset\s+--hard\b/,
    reason: "This will discard ALL uncommitted changes. They cannot be recovered.",
  },
  {
    pattern: /\bgit\s+push\s+.*--force\b/,
    reason: "This will force-push to a remote branch, potentially overwriting others' work.",
  },
  {
    pattern: /\bgit\s+clean\s+-[^\s]*f\b/,
    reason: "This will permanently delete untracked files from your repository.",
  },
  {
    pattern: /\bDROP\s+(TABLE|DATABASE)\b/i,
    reason: "This will permanently delete database tables and all their data.",
  },
  {
    pattern: /\bDELETE\s+FROM\b/i,
    reason: "This will permanently delete rows from a database table.",
  },
  {
    pattern: /\bTRUNCATE\s+(TABLE)?\b/i,
    reason: "This will permanently delete all rows from a database table.",
  },
  {
    pattern: /\bchmod\s+777\b/,
    reason: "This gives everyone full access to the file, which is a security risk.",
  },
  {
    pattern: /\bsudo\s+rm\b/,
    reason: "This permanently deletes files with administrator privileges.",
  },
  {
    pattern: /\bdocker\s+(rm|rmi)\s+/,
    reason: "This will remove a container or image. Make sure you have backups.",
  },
  {
    pattern: /\bnpm\s+publish\b/,
    reason: "This will publish a package to the npm registry. Make sure you're ready.",
  },
];

/**
 * Sets up a safety interceptor that warns before destructive commands.
 *
 * This works by hooking into the terminal shell execution event
 * and showing a confirmation dialog before allowing the command.
 */
export function setupSafetyInterceptor(
  context: vscode.ExtensionContext,
  recorder: CodeForgeRecorder
): void {
  // Hook into terminal shell execution
  context.subscriptions.push(
    vscode.window.onDidStartTerminalShellExecution(async (event) => {
      const commandLine = event.execution.commandLine.value;

      // Check if this command is destructive
      const destructive = DESTRUCTIVE_PATTERNS.find((d) =>
        d.pattern.test(commandLine)
      );

      if (destructive) {
        // Get the safety check from the recorder
        const safety = recorder.checkCommandSafety(commandLine);

        // Show a detailed warning
        const action = await vscode.window.showWarningMessage(
          `CodeForge Safety: ${destructive.reason}`,
          { modal: true },
          "Run Anyway",
          "Cancel",
          "Show Explanation"
        );

        if (action === "Cancel") {
          // Cancel the execution by not continuing
          // Note: VS Code doesn't provide a direct way to cancel terminal execution
          // The user will need to press Ctrl+C in the terminal
          vscode.window.showInformationMessage(
            "CodeForge: Command cancelled. Press Ctrl+C in the terminal if needed."
          );
        } else if (action === "Show Explanation") {
          const panel = vscode.window.createWebviewPanel(
            "codeforgeSafety",
            "CodeForge Safety Check",
            vscode.ViewColumn.Active,
            { enableScripts: false }
          );

          panel.webview.html = `
            <!DOCTYPE html>
            <html>
            <head>
              <style>
                body { font-family: -apple-system, sans-serif; padding: 20px; color: #e0e0e0; background: #1a1a2e; }
                h2 { color: #f77f00; }
                .command { background: #16213e; padding: 10px; border-radius: 6px; font-family: monospace; margin: 10px 0; }
                .explanation { color: #4cc9f0; margin: 10px 0; }
                .warning { color: #e74c3c; font-weight: bold; }
                .fix { color: #2ecc71; margin-top: 10px; }
              </style>
            </head>
            <body>
              <h2>Safety Check</h2>
              <div class="command">${escapeHtml(commandLine)}</div>
              <div class="warning">⚠️ ${escapeHtml(destructive.reason)}</div>
              <div class="explanation">${escapeHtml(safety.explanation)}</div>
              <div class="fix">💡 ${escapeHtml(safety.reason)}</div>
            </body>
            </html>
          `;
        }
        // If "Run Anyway", let it proceed
      }
    })
  );
}

/** Escape HTML special characters */
function escapeHtml(text: string): string {
  return text
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&#039;");
}

File: packages/jetbrains/src/main/kotlin/com/codeforge/SafetyInterceptor.kt
package com.codeforge

import com.intellij.notification.NotificationGroupManager
import com.intellij.notification.NotificationType
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.Messages
import com.intellij.terminal.TerminalWidget
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import javax.swing.Timer

class SafetyInterceptor(private val project: Project) {

    private val logger = Logger.getInstance(SafetyInterceptor::class.java)
    private val manager get() = CodeForgeProjectManager.getInstance(project)

    private val destructivePatterns = listOf(
        DestructivePattern(
            Regex("""\brm\s+(-[rRfF]+\s+|-[a-zA-Z]*[rR][a-zA-Z]*\s+)"""),
            "This will permanently delete files and folders. There is no undo."
        ),
        DestructivePattern(
            Regex("""\bgit\s+reset\s+--hard\b"""),
            "This will discard ALL uncommitted changes. They cannot be recovered."
        ),
        DestructivePattern(
            Regex("""\bgit\s+push\s+.*--force\b"""),
            "This will force-push to a remote branch, potentially overwriting others' work."
        ),
        DestructivePattern(
            Regex("""\bgit\s+clean\s+-[^\s]*f\b"""),
            "This will permanently delete untracked files from your repository."
        ),
        DestructivePattern(
            Regex("""\bDROP\s+(TABLE|DATABASE)\b""", RegexOption.IGNORE_CASE),
            "This will permanently delete database tables and all their data."
        ),
        DestructivePattern(
            Regex("""\bDELETE\s+FROM\b""", RegexOption.IGNORE_CASE),
            "This will permanently delete rows from a database table."
        ),
        DestructivePattern(
            Regex("""\bTRUNCATE\s+(TABLE)?\b""", RegexOption.IGNORE_CASE),
            "This will permanently delete all rows from a database table."
        ),
        DestructivePattern(
            Regex("""\bchmod\s+777\b"""),
            "This gives everyone full access to the file, which is a security risk."
        ),
        DestructivePattern(
            Regex("""\bsudo\s+rm\b"""),
            "This permanently deletes files with administrator privileges."
        ),
        DestructivePattern(
            Regex("""\bdocker\s+(rm|rmi)\s+"""),
            "This will remove a container or image. Make sure you have backups."
        ),
        DestructivePattern(
            Regex("""\bnpm\s+publish\b"""),
            "This will publish a package to the npm registry. Make sure you're ready."
        ),
        DestructivePattern(
            Regex("""\bmvn\s+deploy\b"""),
            "This will deploy artifacts to a Maven repository."
        ),
        DestructivePattern(
            Regex("""\bgradle\s+publish\b"""),
            "This will publish artifacts to a repository."
        )
    )

    fun checkCommand(command: String): SafetyCheckResult {
        val matchedPattern = destructivePatterns.firstOrNull { it.pattern.containsMatchIn(command) }

        if (matchedPattern != null) {
            val safety = manager.checkCommandSafety(command)
            return SafetyCheckResult(
                isDestructive = true,
                reason = matchedPattern.reason,
                level = safety.level,
                explanation = safety.explanation,
                command = command
            )
        }

        return SafetyCheckResult(
            isDestructive = false,
            reason = "",
            level = "safe",
            explanation = manager.checkCommandSafety(command).explanation,
            command = command
        )
    }

    fun showWarningDialog(result: SafetyCheckResult): SafetyDecision {
        if (!result.isDestructive) {
            return SafetyDecision.ALLOW
        }

        val message = buildString {
            appendLine("CodeForge Safety Check")
            appendLine()
            appendLine("Command: ${result.command}")
            appendLine()
            appendLine("Warning: ${result.reason}")
            appendLine()
            appendLine("Explanation: ${result.explanation}")
            appendLine()
            appendLine("Do you want to proceed?")
        }

        val dialogResult = Messages.showYesNoCancelDialog(
            project,
            message,
            "CodeForge Safety Warning",
            "Run Anyway",
            "Cancel",
            "Show Explanation",
            Messages.getWarningIcon()
        )

        return when (dialogResult) {
            Messages.YES -> SafetyDecision.ALLOW
            Messages.NO -> SafetyDecision.CANCEL
            Messages.CANCEL -> SafetyDecision.SHOW_EXPLANATION
            else -> SafetyDecision.CANCEL
        }
    }

    fun showExplanation(result: SafetyCheckResult) {
        val notification = NotificationGroupManager.getInstance()
            .getNotificationGroup("CodeForge Safety")
            .createNotification(
                "CodeForge Safety Check",
                "<b>Command:</b> ${escapeHtml(result.command)}<br>" +
                        "<b>Risk Level:</b> ${result.level.uppercase()}<br>" +
                        "<b>Reason:</b> ${escapeHtml(result.reason)}<br>" +
                        "<b>Explanation:</b> ${escapeHtml(result.explanation)}",
                NotificationType.WARNING
            )
            .notify(project)
    }

    fun showSafetyNotification(result: SafetyCheckResult) {
        val notification = NotificationGroupManager.getInstance()
            .getNotificationGroup("CodeForge Safety")
            .createNotification(
                "CodeForge: Destructive Command Detected",
                "${result.reason}<br><b>Command:</b> ${escapeHtml(result.command)}",
                NotificationType.WARNING
            )
            .notify(project)
    }

    private fun escapeHtml(text: String): String {
        return text
            .replace("&", "&amp;")
            .replace("<", "&lt;")
            .replace(">", "&gt;")
            .replace("\"", "&quot;")
            .replace("'", "&#039;")
    }

    data class DestructivePattern(
        val pattern: Regex,
        val reason: String
    )

    data class SafetyCheckResult(
        val isDestructive: Boolean,
        val reason: String,
        val level: String,
        val explanation: String,
        val command: String
    )

    enum class SafetyDecision {
        ALLOW,
        CANCEL,
        SHOW_EXPLANATION
    }

    companion object {
        private var instance: SafetyInterceptor? = null

        fun getInstance(project: Project): SafetyInterceptor {
            return instance ?: synchronized(this) {
                instance ?: SafetyInterceptor(project).also { instance = it }
            }
        }
    }
}
